Test fails on WSL

The full console output MIGHT be downloadable from [this link](https://temporary.softf1.com/2021/bugs/2021_04_21_14-30-12_ELVM_revision_709bea7963848fca1439268ab97d7eb174f7681b_test_failure_console_output.txt), but the revision is 709bea7963848fca1439268ab97d7eb174f7681b and probably the most interesting console output lines are:

  
    (diff -u out/elc.c.eir.out out/elc.c.eir.tex.out > out/elc.c.eir.tex.out.diff.tmp && mv out/elc.c.eir.tex.out.diff.tmp out/elc.c.eir.tex.out.diff) || (cat out/elc.c.eir.tex.out.diff.tmp ; false)
    --- out/elc.c.eir.out   2021-04-22 08:11:40.740335000 +0300
    +++ out/elc.c.eir.tex.out       2021-04-22 11:34:47.693720200 +0300
    @@ -1,63 +1,2 @@
     === test/elc.in ===
    -var main = function(getchar, putchar) {
    -var a = 0;
    -var b = 0;
    -var c = 0;
    -var d = 0;
    -var bp = 0;
    -var sp = 0;
    -var pc = 0;
    -var mem = new Int32Array(1 << 24);
    -mem[0] = 1;
    -var running = true;
    -
    -var func0 = function() {
    - while (0 <= pc && pc < 512 && running) {
    -  switch (pc) {
    -  case -1:  // dummy
    -   break;
    -
    -  case 0:
    -   if (true) pc = 1 - 1;
    -   break;
    -
    -  case 1:
    -   a = getchar();
    -   if (a == 0) pc = 3 - 1;
    -   break;
    -
    -  case 2:
    -   putchar(a);
    -   if (true) pc = 1 - 1;
    -   break;
    -
    -  case 3:
    -   running = false; break;
    -  }
    -  pc++;
    - }
    -};
    -
    -while (running) {
    - switch (pc / 512 | 0) {
    - case 0:
    -  func0();
    -  break;
    - }
    -}
    -};
    -if (typeof require != 'undefined') {
    - var sys = require('sys');
    - var input = null;
    - var ip = 0;
    - var getchar = function() {
    -  if (input === null)
    -   input = require('fs').readFileSync('/dev/stdin');
    -  return input[ip++] | 0;
    - };
    - var putchar = function(c) {
    -  sys.print(String.fromCharCode(c & 255));
    - };
    - main(getchar, putchar);
    -}
    
    diff.mk:4: recipe for target 'out/elc.c.eir.tex.out.diff' failed
    make: *** [out/elc.c.eir.tex.out.diff] Error 1

I used 

    make -j 3

for building. The environment:

    tsw1@DESKTOP-H12EA8D:~/m_local/bin_p/ELVM/v2021_04_08/elvm$ uname -a
    Linux DESKTOP-H12EA8D 4.4.0-19041-Microsoft #488-Microsoft Mon Sep 01 13:43:00 PST 2020 x86_64 GNU/Linux
    tsw1@DESKTOP-H12EA8D:~/m_local/bin_p/ELVM/v2021_04_08/elvm$ gcc -v
    Using built-in specs.
    COLLECT_GCC=gcc
    COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-linux-gnu/6/lto-wrapper
    Target: x86_64-linux-gnu
    Configured with: ../src/configure -v --with-pkgversion='Debian 6.3.0-18+deb9u1' --with-bugurl=file:///usr/share/doc/gcc-6/README.Bugs --enable-languages=c,ada,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-6 --program-prefix=x86_64-linux-gnu- --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --with-default-libstdcxx-abi=new --enable-gnu-unique-object --disable-vtable-verify --enable-libmpx --enable-plugin --enable-default-pie --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-6-amd64/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-6-amd64 --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-6-amd64 --with-arch-directory=amd64 --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --with-target-system-zlib --enable-objc-gc=auto --enable-multiarch --with-arch-32=i686 --with-abi=m64 --with-multilib-list=m32,m64,mx32 --enable-multilib --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
    Thread model: posix
    gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1)

--- a/Makefile
+++ b/Makefile
@@ -324,9 +324,6 @@ include target.mk
 TARGET := tex
 RUNNER := tools/runtex.sh
 TOOL := tex
-# It seems tex backend does not work for elc anymore.
-# TODO(hamaji): Fix this.
-TEST_FILTER += out/elc.c.eir.tex
 include target.mk
 
 TARGET := cl

--- a/target/tex.c
+++ b/target/tex.c
@@ -15,125 +15,129 @@ static const char* tex_src_str(Inst* inst) {
   return tex_value_str(&inst->src);
 }
 
-static const char* tex_cmp_op_str(Inst* inst) {
-  int op = normalize_cond(inst->op, 0);
-  switch (op) {
-    case JEQ:
-      return "=";
-    case JLE:
-    case JLT:
-      return "<";
-    case JGE:
-    case JGT:
-      return ">";
-    default:
-      error("oops");
-  }
-}
 
 static void tex_emit_inst(Inst* inst) {
   switch (inst->op) {
   case MOV:
-    emit_line("\\edef\\@reg@%s{%s}%%", reg_names[inst->dst.reg], tex_src_str(inst));
+    emit_line("\\@op@mov\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg], tex_src_str(inst));
     break;
 
   case ADD:
-    emit_line("\\count0=\\@reg@%s\\relax", reg_names[inst->dst.reg]);
-    emit_line("\\advance\\count0by%s\\relax", tex_src_str(inst));
-    emit_line("\\ifnum\\count0>%d\\advance\\count0by-%s\\fi", UINT_MAX, "16777216");
-    emit_line("\\edef\\@reg@%s{\\the\\count0}%%", reg_names[inst->dst.reg]);
+    emit_line("\\@op@add\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg], tex_src_str(inst));
     break;
 
   case SUB:
-    emit_line("\\count0=\\@reg@%s\\relax", reg_names[inst->dst.reg]);
-    emit_line("\\advance\\count0by-%s\\relax", tex_src_str(inst));
-    emit_line("\\ifnum\\count0<0\\advance\\count0by%s\\fi", "16777216");
-    emit_line("\\edef\\@reg@%s{\\the\\count0}%%", reg_names[inst->dst.reg]);
+    emit_line("\\@op@sub\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg], tex_src_str(inst));
     break;
 
   case LOAD:
-    emit_line("\\def\\@reg@%s{0}%%", reg_names[inst->dst.reg]);
-    emit_line("\\expandafter\\ifx\\csname @mem@%s\\endcsname\\relax\\else\\edef\\@reg@%s{\\csname @mem@%s\\endcsname}\\fi",
-              tex_src_str(inst), reg_names[inst->dst.reg], tex_src_str(inst));
+    emit_line("\\@op@load\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg], tex_src_str(inst));
     break;
 
   case STORE:
-    emit_line("\\expandafter\\let\\csname @mem@%s\\endcsname\\@reg@%s",
-              tex_src_str(inst), reg_names[inst->dst.reg]);
+    emit_line("\\@op@store\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg], tex_src_str(inst));
     break;
 
   case PUTC:
-    // a output char c is represented by C where C is the code of c.
-    emit_line("\\immediate\\write\\@out{%s}%%", tex_src_str(inst));
+    emit_line("\\@op@putc{%s}%%", tex_src_str(inst));
     break;
 
   case GETC:
-    // input is also...
-    emit_line("\\read-1to\\@temp\\count0=\\@temp\\edef\\@reg@%s{\\the\\count0}%%", reg_names[inst->dst.reg]);
+    emit_line("\\@op@getc\\@reg@%s", reg_names[inst->dst.reg]);
     break;
 
   case EXIT:
-    emit_line("\\let\\@@next\\relax");
+    emit_line("\\@op@exit");
     break;
 
   case DUMP:
     break;
 
   case EQ:
+    emit_line("\\@op@eq\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg],
+              tex_src_str(inst));
+    break;
+
   case LT:
+    emit_line("\\@op@lt\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg],
+              tex_src_str(inst));
+    break;
+
   case GT:
-    emit_line("\\ifnum\\@reg@%s%s%s\\edef\\@reg@%s{1}\\else\\edef\\@reg@%s{0}\\fi%%",
+    emit_line("\\@op@gt\\@reg@%s{%s}%%",
               reg_names[inst->dst.reg],
-              tex_cmp_op_str(inst),
-              tex_src_str(inst),
-              reg_names[inst->dst.reg],
-              reg_names[inst->dst.reg]);
+              tex_src_str(inst));
     break;
+
   case NE:
-    emit_line("\\ifnum\\@reg@%s=%s\\edef\\@reg@%s{0}\\else\\edef\\@reg@%s{1}\\fi%%",
-              reg_names[inst->dst.reg], tex_src_str(inst),
-              reg_names[inst->dst.reg], reg_names[inst->dst.reg]);
+    emit_line("\\@op@ne\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg],
+              tex_src_str(inst));
     break;
+
   case LE:
+    emit_line("\\@op@le\\@reg@%s{%s}%%",
+              reg_names[inst->dst.reg],
+              tex_src_str(inst));
+    break;
+
   case GE:
-    emit_line("\\ifnum\\@reg@%s=%s\\edef\\@reg@%s{1}\\else"
-              "\\ifnum\\@reg@%s%s%s\\edef\\@reg@%s{1}\\else\\edef\\@reg@%s{0}\\fi\\fi%%",
-              reg_names[inst->dst.reg], tex_src_str(inst),
+    emit_line("\\@op@ge\\@reg@%s{%s}%%",
               reg_names[inst->dst.reg],
-              reg_names[inst->dst.reg], tex_cmp_op_str(inst), tex_src_str(inst),
-              reg_names[inst->dst.reg],
-              reg_names[inst->dst.reg]);
+              tex_src_str(inst));
     break;
 
   case JEQ:
+    emit_line("\\@op@jeq\\@reg@%s{%s}{%s}%%",
+              reg_names[inst->dst.reg],
+              tex_src_str(inst),
+              tex_value_str(&inst->jmp));
+    break;
+
   case JLT:
+    emit_line("\\@op@jlt\\@reg@%s{%s}{%s}%%",
+              reg_names[inst->dst.reg],
+              tex_src_str(inst),
+              tex_value_str(&inst->jmp));
+    break;
+
   case JGT:
-    emit_line("\\count0=%s\\relax\\advance\\count0by-1\\relax", tex_value_str(&inst->jmp));
-    emit_line("\\ifnum\\@reg@%s%s%s\\edef\\@reg@pc{\\the\\count0}\\fi",
+    emit_line("\\@op@jgt\\@reg@%s{%s}{%s}%%",
               reg_names[inst->dst.reg],
-              tex_cmp_op_str(inst),
-              tex_src_str(inst));
+              tex_src_str(inst),
+              tex_value_str(&inst->jmp));
     break;
+
   case JNE:
-    emit_line("\\count0=%s\\relax\\advance\\count0by-1\\relax", tex_value_str(&inst->jmp));
-    emit_line("\\ifnum\\@reg@%s=%s\\else\\edef\\@reg@pc{\\the\\count0}\\fi",
+    emit_line("\\@op@jne\\@reg@%s{%s}{%s}%%",
               reg_names[inst->dst.reg],
-              tex_src_str(inst));
+              tex_src_str(inst),
+              tex_value_str(&inst->jmp));
     break;
+
   case JLE:
-  case JGE:
-    emit_line("\\count0=%s\\relax\\advance\\count0by-1\\relax", tex_value_str(&inst->jmp));
-    emit_line("\\ifnum\\@reg@%s=%s\\edef\\@reg@pc{\\the\\count0}\\fi",
+    emit_line("\\@op@jle\\@reg@%s{%s}{%s}%%",
               reg_names[inst->dst.reg],
-              tex_src_str(inst));
-    emit_line("\\ifnum\\@reg@%s%s%s\\edef\\@reg@pc{\\the\\count0}\\fi",
-              reg_names[inst->dst.reg],
-              tex_cmp_op_str(inst),
-              tex_src_str(inst));
+              tex_src_str(inst),
+              tex_value_str(&inst->jmp));
     break;
+
+  case JGE:
+    emit_line("\\@op@jge\\@reg@%s{%s}{%s}%%",
+              reg_names[inst->dst.reg],
+              tex_src_str(inst),
+              tex_value_str(&inst->jmp));
+    break;
+
   case JMP:
-    emit_line("\\count0=%s\\relax\\advance\\count0by-1\\relax", tex_value_str(&inst->jmp));
-    emit_line("\\edef\\@reg@pc{\\the\\count0}%%");
+    emit_line("\\@op@jmp{%s}%%", tex_value_str(&inst->jmp));
     break;
 
   default:
@@ -158,6 +162,59 @@ static void tex_init_state(Data* data) {
 
 void target_tex(Module* module) {
   emit_line("\\catcode`\\@=11\\relax");
+
+  // Define macros for EIR's operations
+  emit_line("\\def\\@op@mov#1#2{\\edef#1{#2}}");
+  emit_line("\\def\\@op@add#1#2{"
+            "\\count0=#1\\relax\\advance\\count0by#2\\relax"
+            "\\ifnum\\count0>%d\\advance\\count0by-%s\\fi"
+            "\\edef#1{\\the\\count0}}",
+            UINT_MAX,
+            "16777216");
+  emit_line("\\def\\@op@sub#1#2{"
+            "\\count0=#1\\relax\\advance\\count0by-#2\\relax"
+            "\\ifnum\\count0<0\\advance\\count0by%s\\fi"
+            "\\edef#1{\\the\\count0}}", "16777216");
+  emit_line("\\def\\@op@load#1#2{"
+            "\\edef#1{"
+            "\\expandafter\\ifx\\csname @mem@#2\\endcsname\\relax0"
+            "\\else\\csname @mem@#2\\endcsname\\fi}}");
+  emit_line("\\def\\@op@store#1#2{"
+            "\\expandafter\\let\\csname @mem@#2\\endcsname#1}");
+  emit_line("\\def\\@op@putc#1{\\immediate\\write\\@out{#1}}");
+  emit_line("\\def\\@op@getc#1{"
+            "\\read-1to\\@temp\\count0=\\@temp\\edef#1{\\the\\count0}}");
+  emit_line("\\def\\@op@exit{\\let\\@@next\\relax}");
+  emit_line("\\def\\@op@eq#1#2{\\ifnum#1=#2\\edef#1{1}\\else\\edef#1{0}\\fi}");
+  emit_line("\\def\\@op@lt#1#2{\\ifnum#1<#2\\edef#1{1}\\else\\edef#1{0}\\fi}");
+  emit_line("\\def\\@op@gt#1#2{\\ifnum#1>#2\\edef#1{1}\\else\\edef#1{0}\\fi}");
+  emit_line("\\def\\@op@ne#1#2{\\ifnum#1=#2\\edef#1{0}\\else\\edef#1{1}\\fi}");
+  emit_line("\\def\\@op@le#1#2{"
+            "\\ifnum#1=#2\\edef#1{1}"
+            "\\else\\ifnum#1<#2\\edef#1{1}"
+            "\\else\\edef#1{0}\\fi\\fi}");
+  emit_line("\\def\\@op@ge#1#2{"
+            "\\ifnum#1=#2\\edef#1{1}"
+            "\\else\\ifnum#1>#2\\edef#1{1}"
+            "\\else\\edef#1{0}\\fi\\fi}");
+  emit_line("\\def\\@op@jmp@common#1{"
+            "\\count0=#1\\relax\\advance\\count0by-1\\relax"
+            "\\edef\\@reg@pc{\\the\\count0}}");
+  emit_line("\\def\\@op@jeq#1#2#3{"
+            "\\ifnum#1=#2\\relax\\@op@jmp@common{#3}\\fi}");
+  emit_line("\\def\\@op@jlt#1#2#3{"
+            "\\ifnum#1<#2\\relax\\@op@jmp@common{#3}\\fi}");
+  emit_line("\\def\\@op@jgt#1#2#3{"
+            "\\ifnum#1>#2\\relax\\@op@jmp@common{#3}\\fi}");
+  emit_line("\\def\\@op@jne#1#2#3{"
+            "\\ifnum#1=#2\\else\\relax\\@op@jmp@common{#3}\\fi}");
+  emit_line("\\def\\@op@jle#1#2#3{"
+            "\\ifnum#1=#2\\relax\\@op@jmp@common{#3}"
+            "\\else\\ifnum#1<#2\\relax\\@op@jmp@common{#3}\\fi\\fi}");
+  emit_line("\\def\\@op@jge#1#2#3{"
+            "\\ifnum#1=#2\\relax\\@op@jmp@common{#3}"
+            "\\else\\ifnum#1>#2\\relax\\@op@jmp@common{#3}\\fi\\fi}");
+  emit_line("\\let\\@op@jmp\\@op@jmp@common");
 
   // Initialize register and memory.
   tex_init_state(module->data);

